src/app/modules/user/user.service.ts

```diff
@@ -148,8 +150,38 @@ export const updateUserContact = async (
 export const getPopulatedUserById = async (
   userId: IUserSchema['_id'],
 ): Promise<IPopulatedUser> => {
-  return User.findById(userId).populate({
+  return UserModel.findById(userId).populate({
     path: 'agency',
     model: AGENCY_SCHEMA_ID,
   })
 }
+
+/**
+ * Updates or creates a user document with the given email and return the user.
+ * @param email the email of the user to retrieve
+ * @param agency the agency document to associate with the user
+ * @returns the upserted user document
+ * @throws {InvalidDomainError} on invalid email
+ * @throws {Error} on upsert failure
+ */
+export const upsertAndReturnUser = async (email: string, agency: IAgency) => {
```

Comment:
may I suggest renaming this function to a business purpose oriented name e.g. `createNewUser`, then moving the upsert operations to the `UserModel` instead, as per what you've done with `TokenSchema.statics.upsertOtp`?